subsys/net/lib/dns/dispatcher.c

```diff
@@ -0,0 +1,304 @@
+
+	/* Make sure that we can read DNS id, flags and rcode */
+	if (dns_msg.msg_size < (sizeof(uint16_t) + sizeof(uint16_t))) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	if (dns_header_rcode(dns_msg.msg) == DNS_HEADER_REFUSED) {
+		ret = -EINVAL;
+		goto done;
+	}
+
+	is_query = (dns_header_qr(dns_msg.msg) == DNS_QUERY);
+	if (is_query) {
+		if (dispatcher->type == DNS_SOCKET_RESPONDER) {
+			/* Call the responder callback */
+			ret = dispatcher->cb(dispatcher->ctx, sock,
+					     addr, addrlen,
+					     dns_data, data_len);
+		} else if (dispatcher->pair) {
+			if (dispatcher->pair->type == DNS_SOCKET_RESPONDER) {
+				ret = dispatcher->pair->cb(dispatcher->pair->ctx, sock,
+							   addr, addrlen,
+							   dns_data, data_len);
+			} else {
+				/* The pair must be a resolver */
+				CHECKIF(dispatcher->pair->type != DNS_SOCKET_RESOLVER) {
+					NET_DBG("Duplicate dispatcher type found (%d)",
+						dispatcher->pair->type);
+				}
+
+				ret = -EINVAL;
+			}
+		} else {
+			/* Discard the message as it was a query and there are none
+			 * expecting a query.
+			 */
+			ret = -ENOENT;
+		}
+	} else {
+		/* So this was an answer to a query that was made by resolver.
+		 */
+		if (dispatcher->type == DNS_SOCKET_RESOLVER) {
+			/* Call the resolver callback */
+			ret = dispatcher->cb(dispatcher->ctx, sock,
+					     addr, addrlen,
+					     dns_data, data_len);
+		} else if (dispatcher->pair) {
+			if (dispatcher->pair->type == DNS_SOCKET_RESOLVER) {
+				ret = dispatcher->pair->cb(dispatcher->pair->ctx, sock,
+							   addr, addrlen,
+							   dns_data, data_len);
+			} else {
+				/* The pair must be a responder */
+				CHECKIF(dispatcher->pair->type != DNS_SOCKET_RESPONDER) {
+					NET_DBG("Duplicate dispatcher type found (%d)",
+						dispatcher->pair->type);
+				}
+
+				ret = -EINVAL;
+			}
+		} else {
+			/* Discard the message as it was a query and there are none
+			 * expecting a query.
+			 */
+			ret = -ENOENT;
+		}
+	}
+
+done:
+	return ret;
+}
+
+static int recv_data(struct net_socket_service_event *pev)
+{
+	struct dns_socket_dispatcher *dispatcher = pev->user_data;
+	socklen_t optlen = sizeof(int);
+	struct net_buf *dns_data = NULL;
+	struct sockaddr addr;
+	size_t addrlen;
+	int family, sock_error;
+	int ret = 0, len;
+
+	k_mutex_lock(&dispatcher->lock, K_FOREVER);
+
+	(void)zsock_getsockopt(pev->event.fd, SOL_SOCKET,
+			       SO_DOMAIN, &family, &optlen);
+
+	if ((pev->event.revents & ZSOCK_POLLERR) ||
+	    (pev->event.revents & ZSOCK_POLLNVAL)) {
+		(void)zsock_getsockopt(pev->event.fd, SOL_SOCKET,
+				       SO_ERROR, &sock_error, &optlen);
+		NET_ERR("Receiver IPv%d socket error (%d)",
+			family == AF_INET ? 4 : 6, sock_error);
+		ret = DNS_EAI_SYSTEM;
+		goto unlock;
+	}
+
+	addrlen = (family == AF_INET) ? sizeof(struct sockaddr_in) :
+		sizeof(struct sockaddr_in6);
+
+	dns_data = net_buf_alloc(&dns_msg_pool, dispatcher->buf_timeout);
+	if (!dns_data) {
+		ret = DNS_EAI_MEMORY;
+		goto free_buf;
```

Comment:
It's rather harmless given there's a pointer check under `free_buf`, but since the allocation failed we could jump directly to `unlock` label here.